cmd/snap-device-helper/snap-device-helper.c

```diff
@@ -43,6 +43,21 @@ static unsigned long must_strtoul(const char *str) {
     return val;
 }
 
+static void reverse_component_separator_encoding(char *tag, const char *original) {
+    char *separator = strstr(tag, "__");
+    if (separator == NULL) {
+        return;
+    }
+
+    // if there is another double underscore anywhere in the string, something is wrong
+    if (strstr(separator + 2, "__") != NULL) {
+        die("malformed tag \"%s\"", original);
+    }
+
+    *separator = '+';
+    memmove(separator + 1, separator + 2, strlen(separator + 2) + 1);
```

Comment:
Hmm, is the size okay? I was expecting us to move strlen(separator+1) as we are shortening the string by one byte.

Reply:
Wouldn't `(strlen(separator + 2) + 1) == strlen(separator + 1)`? Although I think the former is more clear, since we want to copy everything after the separator, including the null terminator.

Reply:
Apologies, I think this is okay.